libs/ledgerjs/packages/hw-transport-react-native-ble/android/src/main/java/com/hwtransportreactnativeble/tasks/Queue.kt

```diff
@@ -0,0 +1,218 @@
+                    val rawItem = rawTasks.getJSONObject(i)
+                    newTasks.add(
+                        Item(
+                            id = rawItem.getInt("id"),
+                            operation = rawItem.getString("operation"),
+                            appName = rawItem.getString("appName")
+                        )
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            onErrorWrapper(e)
+            return
+        }
+
+
+        for (call in client.dispatcher.queuedCalls()) call.cancel()
+        val body: RequestBody = rawQueue.toRequestBody(JSON)
+
+        val request = Request.Builder()
+            .url(packQueueURL)
+            .method("PUT", body)
+            .build()
+
+        client.newCall(request).enqueue(object : Callback {
+            override fun onFailure(call: Call, e: IOException) {
+                onErrorWrapper(IOException("Network error")) // FIXME, be more precise?
+            }
+
+            override fun onResponse(call: Call, response: Response) {
+                response.use {
+                    if (!response.isSuccessful) {
+                        onErrorWrapper(IOException("Unexpected code $response"))
+                    } else {
+                        token = response.body?.string()
+                        tasks.clear()
+                        tasks.addAll(newTasks)
+
+                        if (autoStart) {
+                            startRunner()
+                        }
+                    }
+
+                }
+            }
+        })
+    }
+
+    private fun startRunner() {
+        if (index >= tasks.size) return
+        val item = tasks[index]
+        Timber.d("Starting runner for ${item.toString()}")
+
+        eventEmitter.dispatch(Arguments.createMap().apply {
+            putString("event", "task")
+            putString("type", RunnerAction.runStart.toString())
+            putMap("data", Arguments.createMap().apply {
+                putString("name", item.appName)
+                putString("type", item.operation)
+            })
+        })
+
+        runner = Runner(
+            endpoint = websocketURL,
+            initialMessage = "{\"token\":\"$token\",\"index\":$index}",
+            bleManager = bleManager,
+
+            onDone = ::onDoneWrapper,
+            onEvent = ::onEventWrapper,
+            onStop = ::none
+        )
+    }
+    fun none() {
+
+    }
+
+    fun stop() {
+        isStopped = true
+        token = ""
+        index = 0
+        item = null
+        tasks.clear()
+        for (call in client.dispatcher.queuedCalls()) call.cancel()
+
+        runner?.stop()
+        runner = null
+    }
+
+    private fun onDoneWrapper(doneMessage: String) {
+        if (index >= tasks.size) return
+        val item = tasks[index]
+        Timber.d("Completed runner for ${item.toString()}")
+
+        eventEmitter.dispatch(Arguments.createMap().apply {
```

Comment:
Technically `onEventWrapper(RunnerAction.runSuccess, Arguments.createMap().apply {})` should be enough since I'm polyfilling the data with exactly name and type in the wrapper, good catch